crates/typst/src/layout/grid/layout.rs

```diff
@@ -216,38 +226,123 @@ impl CellGrid {
         inset: Sides<Rel<Length>>,
         engine: &mut Engine,
         styles: StyleChain,
+        span: Span,
     ) -> SourceResult<Self> {
         // Number of content columns: Always at least one.
         let c = tracks.x.len().max(1);
 
-        // If not all columns in the last row have cells, we will add empty
-        // cells and complete the row so that those positions are susceptible
-        // to show rules and receive grid styling.
-        // We apply '% c' twice so that 'cells_remaining' is zero when
-        // the last row is already filled (then 'cell_count % c' would be zero).
+        // We can't just use the cell's index in the 'cells' vector to
+        // determine its automatic position, since cells could have arbitrary
+        // positions, so the position of a cell in 'cells' can differ from its
+        // final position in 'resolved_cells' (see below).
+        // Therefore, we use a counter, 'auto_index', to determine the position
+        // of the next cell with (x: auto, y: auto). It is only stepped when
+        // a cell with (x: auto, y: auto), usually the vast majority, is found.
+        let mut auto_index = 0;
+
+        // We have to rebuild the grid to account for arbitrary positions.
+        // Create at least 'cells.len()' positions, since there will be at
+        // least 'cells.len()' cells, even though some of them might be placed
+        // in arbitrary positions and thus cause the grid to expand.
         let cell_count = cells.len();
-        let cells_remaining = (c - cell_count % c) % c;
-        let cells = cells
-            .iter()
-            .cloned()
-            .chain(std::iter::repeat_with(T::default).take(cells_remaining))
+        let mut resolved_cells: Vec<Option<Cell>> = Vec::with_capacity(cell_count);
+        for cell in cells.iter().cloned() {
+            // Let's calculate the cell's final position based on its
+            // requested position.
+            let resolved_index = {
+                let cell_x = cell.x(styles);
+                let cell_y = cell.y(styles);
+                resolve_cell_position(cell_x, cell_y, &resolved_cells, &mut auto_index, c)
+                    .at(span)?
+            };
+            let x = resolved_index % c;
+            let y = resolved_index / c;
+
+            // Let's resolve the cell so it can determine its own fields
+            // based on its final position.
+            let cell = cell.resolve_cell(
+                x,
+                y,
+                &fill.resolve(engine, x, y)?,
+                align.resolve(engine, x, y)?,
+                inset,
+                styles,
+            );
+
+            // Now let's check if the cell's position is valid.
+            if let Some(existing_cell) = resolved_cells.get_mut(resolved_index) {
```

Comment:
I think we can simplify this case distinction a bit by doing the following:

1. `if resolved_index >= resolved_cells.len() { resolved_cells.resize(resolved_index + 1, None) }`
2. `let slot = &mut resolved_cells[resolved_index];`
3. Check for existing cell
4. Assign cell

resize + assign is perhaps a tiny bit slower than push but on the upside we have a single code path and less potential for off-by-one bugs.

Reply:
I based the current design on the principle of having the common use case (no positioning) be as fast as possible, but perhaps this could make sense if combined with the idea of pushing full rows. Will experiment a bit.

Reply:
Seems to have mostly equivalent performance, so should be fine. Implemented.